pkg/container/priority_queue.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package container
+
+import (
+	"container/heap"
+
+	"github.com/cilium/hubble/pkg/api/v1"
+)
+
+// PriorityQueue is a priority queue of events that implements heap.Interface.
+// Events are ordered based on their timestamp value, the oldest event having
+// the higher priority.
+//
+// This implementation is loosely based on the priority queue example in
+// "container/heap".
+type PriorityQueue struct {
+	pq priorityQueue
+}
+type priorityQueue []*v1.Event
+
+// Ensure that priorityQueue implements heap.Interface.
+var _ heap.Interface = (*priorityQueue)(nil)
+
+// NewPriorityQueue creates a new PriorityQueue with initial capacity n.
+// It may optionally be initialized with events in which case the initial
+// capacity is set to n or the number of events, whichever is greater.
+func NewPriorityQueue(n int, events ...*v1.Event) *PriorityQueue {
```

Comment:
i woudln't commit varargs to the list of events, i'd either use the options pattern (and a list of events as one of the options), or just omit it all together

Reply:
Using the option pattern seems overkill for this use-case. However, I have nothing against omitting it altogether as, all things considered, I don't think it's really required/useful anyway.